### PR TITLE
Floor Refactor

### PR DESCRIPTION
**Pull request by connorpoland**
the floor subsystem now listens on 1 socket the way god intended. Depending on the class of message recieved the data is given to supporting classes ArrivalSensor, TripTracker, and Floor respectively. Follow the README for an indepth explination on the responsabilities of each of these supporting classes.

### CODE REVIEW

**Comment by vikrambombhi** [comment]
Is this ready to be tested?

File: src/floor/FloorSystem.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+					//simulate the passenger pressing their destination
+					tripTracker.sendTripDestination(arrivingElevator, arrivingFloor, Direction.DOWN);
+					
+					//modify model
+					floors[arrivingFloor].setDownLamp(false);
+					System.out.println("Floor System: Passenger going DOWN stepped into elevator "+arrivingElevator);
+				}
+		}
+		//keep listening forever...
+		listenAndDispatch();
```
Comment (JackyChiu) [inline src/floor/FloorSystem.java]: Does Java have tail call optimization? This probably won't matter but if we run the process long enough these stack frame's memory are going to build up

File: src/floor/TripTracker.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import java.net.SocketException;
+
+import messages.ElevatorRequestMessage.Direction;
+import messages.FloorRequestMessage;
+import messages.Message;
+
+public class TripTracker {
+
+	DatagramSocket sendSocket;
+	DatagramPacket sendPacket;
```
Comment (JackyChiu) [inline src/floor/TripTracker.java]: Nitpick: Instead of using an ivar sendPacket, initialize it here and use the Message.send

File: src/floor/RequestSimulator.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import java.net.SocketException;
+
+import messages.ElevatorRequestMessage;
+import messages.ElevatorRequestMessage.Direction;
+import messages.Message;
+
+public class RequestSimulator implements Runnable{
+	
+	DatagramSocket sendSocket;
+	DatagramPacket sendPacket;
```
Comment (JackyChiu) [inline src/floor/RequestSimulator.java]: Nitpick: Instead of using an ivar sendPacket, initialize it here and use the Message.send

File: src/messages/FloorArrivalMessage.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+public class FloorArrivalMessage implements Message, Serializable {
+
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = -4085784915939113686L;
+	
+	private int elevator;
+	private int floor;
+	public enum Direction { UP, DOWN };
```
Comment (JackyChiu) [inline src/messages/FloorArrivalMessage.java]: We should make just one `Direction` enum in the `messages` pkg for all messages

File: src/floor/FloorSystem.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+import java.net.SocketException;
+
+import floor.Floor.directionLampState;
+import messages.ElevatorRequestMessage;
+import messages.ElevatorRequestMessage.Direction;
+import messages.FloorArrivalMessage;
+import messages.FloorTravelMessage;
+import messages.Message;
+
+public class FloorSystem {
```
Comment (JackyChiu) [inline src/floor/FloorSystem.java]: This class feels a bit odd to me because we're supposed to have a one-floor subsystem per floor, but this class is modeling the whole thing. Do you have a reason for this decision?

File: src/floor/RequestSimulator.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+			Thread.sleep(1000);
+		} catch (InterruptedException e) {
+			e.printStackTrace();
+		}
+		
+		ElevatorRequestMessage m = new ElevatorRequestMessage();
+		m.setDirection(Direction.UP);
+		m.setFloor(1);
+		byte[] data = Message.serialize(m);
+		//inform the floor subsystem
```
Comment (JackyChiu) [inline src/floor/RequestSimulator.java]: We're supposed to expect that the message goes to the correct floor subsystem right? Not one big sub system?

File: src/floor/ArrivalSensor.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import java.net.DatagramSocket;
+import java.net.SocketException;
+
+import messages.FloorArrivalMessage;
+import messages.FloorArrivalMessage.Direction;
+import messages.Message;
+
+public class ArrivalSensor extends Thread{
+	
+	DatagramPacket sendPacket;
```
Comment (kirinrastogi) [inline src/floor/ArrivalSensor.java]: can we remove the packet and pass it around as a parameter?

File: src/floor/Floor.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
+package floor;
+
+public class Floor {
+	
+	//static model
+	int floorNum;
```
Comment (kirinrastogi) [inline src/floor/Floor.java]: can you make these all `private` so we don't lose styling marks?

File: src/floor/ArrivalSensor.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	DatagramSocket sendSocket;
+	
+	int elevatorShaft;
```
Comment (kirinrastogi) [inline src/floor/ArrivalSensor.java]: `private` pls...

File: src/floor/FloorSystem.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+		System.out.println("Floor System: Beginning simulation");
+		requestSimulator.start();
+		listenAndDispatch();
+	}
+	
+	public void listenAndDispatch() {	
+		
+		//listen for incoming messages
+		byte[] data = new byte[100];
+		Message m = Message.deserialize(Message.receive(receiveSocket, data).getData());
```
Comment (kirinrastogi) [inline src/floor/FloorSystem.java]: I changed the signature of receive.

File: src/floor/Floor.java (hunk 8 above)
Comment (kirinrastogi) [reply src/floor/Floor.java]: unless the spec says it doesn't matter

File: src/floor/FloorSystem.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+public class FloorSystem {
+
+	DatagramSocket receiveSocket;
+	
+	//model of the system
+	Floor[] floors;
+	
+	//this has to be a thread because it has to sleep while the rest of the simulation is active
+	Thread[] arrivalSensors;
```
Comment (kirinrastogi) [inline src/floor/FloorSystem.java]: whats the reason these are Threads and not ArrivalSensors?

File: src/floor/ArrivalSensor.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.SocketException;
+
+import messages.FloorArrivalMessage;
+import messages.FloorArrivalMessage.Direction;
+import messages.Message;
+
+public class ArrivalSensor extends Thread{
```
Comment (kirinrastogi) [inline src/floor/ArrivalSensor.java]: This was `implements Runnable` before... Why the switch?

File: src/floor/FloorSystem.java (hunk 1 above)
Comment (kirinrastogi) [reply src/floor/FloorSystem.java]: it does not, this should be changed. https://softwareengineering.stackexchange.com/questions/272061/why-doesnt-java-have-optimization-for-tail-recursion-at-all

tl;dr things guess stack pointer locations by counting stack frames & amount of parameters in subframes
